src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -0,0 +1,120 @@
+            context = managedIdentityContext(request, petName)
+            azureUami <- IO(manager.identities().define(petName.value)
+              .withRegion(managedIdentityRegion)
+              .withExistingResourceGroup(request.managedResourceGroupName.value)
+              .withTags(managedIdentityTags(user).asJava)
+              .create(context))
+            petToCreate = PetManagedIdentity(id, ManagedIdentityObjectId(azureUami.id()), ManagedIdentityDisplayName(azureUami.name()))
+            createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
+          } yield (createdPet, true)
+      }
+    } yield pet
+  }
+
+  /**
+    * Resolves a managed resource group in Azure and returns the terra.billingProfileId tag value.
+    * This is used for access control checks.
+    */
+  def getBillingProfileId(request: GetOrCreatePetManagedIdentityRequest): IO[Option[ResourceId]] = {
+    for {
+      resourceManager <- crlService.buildResourceManager(request.tenantId, request.subscriptionId)
+      mrg <- IO(resourceManager.resourceGroups().getByName(request.managedResourceGroupName.value)).attempt
+      billingProfileOpt = mrg.toOption.flatMap(_.tags().asScala.get(billingProfileTag))
+    } yield billingProfileOpt.map(ResourceId(_))
+  }
+
+  /**
+    * Validates a managed resource group. Algorithm:
+    * 1. Resolve the MRG as the publisher
+    * 2. Get the managed app id from the MRG
+    * 3. Resolve the managed app as the publisher
+    * 4. Get the managed app "plan" id
+    * 4. Validate the plan id matches the configured value
+    */
+  private def validateManagedResourceGroup(request: GetOrCreatePetManagedIdentityRequest): IO[Unit] = {
+    for {
+      resourceManager <- crlService.buildResourceManager(request.tenantId, request.subscriptionId)
+      resolvedMrg <- IO(resourceManager.resourceGroups().getByName(request.managedResourceGroupName.value)).attempt
+      managedByOpt = for {
+        mrg <- resolvedMrg.toOption
+        im <- Option(mrg.innerModel())
+        mb <- Option(im.managedBy())
+      } yield mb
+      managedBy <- IO.fromOption(managedByOpt)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, s"Validation failed: could not retrieve managed app for managed resource group ${request.managedResourceGroupName.value}"))
+      )
+      resolvedManagedApp <- IO(resourceManager.genericResources().getById(managedBy)).attempt
+      planIdOpt = for {
+        ma <- resolvedManagedApp.toOption
+        p <- Option(ma.plan())
+      } yield p.name()
+      planId <- IO.fromOption(planIdOpt)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "Validation failed: could not retrieve plan for managed app"))
+      )
+      _ <- IO.raiseUnless(planId == crlService.getManagedAppPlanId)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, "Validation failed: wrong managed app plan"))
+      )
+    } yield ()
```

Comment:
Could this be simplified with helper functions? Nested for/yields aren't my favorite.

Reply:
Factored these out into helper functions.